server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -378,9 +377,22 @@ protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucket
                     // the filter doesn't match any docs
                     continue;
                 }
-                TotalHitCountCollector collector = new TotalHitCountCollector();
-                scorer.score(collector, live);
-                incrementBucketDocCount(filterOrd, collector.getTotalHits());
+                if (sub == LeafBucketCollector.NO_OP_COLLECTOR) {
+                    TotalHitCountCollector collector = new TotalHitCountCollector();
+                    scorer.score(collector, live);
+                    incrementBucketDocCount(filterOrd, collector.getTotalHits());
+                } else {
+                    /*
+                     * We can use the pre-constructed leaf collected for the first
+                     * filter. But it almost certainly not going to work for the
+                     * second one because it'll try to "go backwards". So we build
+                     * a new one for each subsequent filter.
+                     */
+                    LeafBucketCollector filterLeafCollector = filterOrd == 0 ? sub : collectableSubAggregators.getLeafCollector(ctx);
+                    SubCollector collector = new SubCollector(filterOrd, filterLeafCollector);
+                    scorer.score(collector, live);
+                    incrementBucketDocCount(filterOrd, collector.total);
+                }
```

Comment:
I think it's possible that this could actually be slower than the standard execution mechanism. I wonder if we need an escape hatch so folks can dodge this mechanism if it proves a bad idea.


Also: there is another possible implementation here that involves collecting a block of matches for each filter and then running all of the children in parallel. I'm not sure if it'll be faster or not. It kind of depends on the speed of iterating the doc values. It is a little more complex so I didn't do it.

Reply:
After taking a couple of days away I think the block matching mechanism is probably better here. Its much simpler to estimate the cost. Also - I'd love to know why the old way is so slow - the block based mechanism feels like it'd be fast and it reads quite similarly to the Compatible mechanism. I *think* the big difference is that we don't get to join the main query with the filter query. So it can't skip matches effectively. Maybe. I've got to play.